### ISSUE TITLE
Missing "created" invalidates signature? 

### ISSUE THREAD

**Issue by cyberphone** [w3c-dvcg/ld-signatures#5]
If I got this right, the absence of a "created" property requires the verifier to create an artificial ditto holding the current time.  Either "created" is unsigned (which would be strange), or validation will fail.

**Comment by msporny** [comment]
Yes, missing "created" invalidates the signature per the current specification. Also, anything placed in the "signature" block is signed per the newest algorithm (unless the @context doesn't specify the value... we expect future versions of the algorithm to throw an error if a value is dropped during signing unless explicitly stated by the developer that doing so is OK). 

We did play around with the notion that "created" was optional. At present, it represents a form of pseudo-nonce... it adds information to the signature that enables a receiver to determine if they want to process the signature at all. For example, some verifiers may decide that they don't want to process any signature that is more than 15 minutes old.

We're still open to making "created" optional, but we would need a very good use case where requiring "created" is problematic for a particular application. The best use case we've heard of so far, which is still a weak argument, is that the signer may not have access to a good/valid time source. The counter argument to that has been that devices that don't have access to a good/valid time source shouldn't be creating or verifying signatures.

**Comment by cyberphone** [comment]
This touches a generic issue with LDS which is that it wants to do more than just sign.  I'm sure it is a perfect fit for many applications but as a general purpose solution this notion (IMO) creates more problems than it solves.  As an example most real-world messages (signed or unsigned), include some kind of timestamp.  Designers would presumably like this to be in the same place and having the same name in a given scheme.  I'm personally quite picky and can spend _days_ thinking about the "best" name for a property :-)

FWIW, I have recently upgraded JCS to deal with signature-related meta-data in a _dynamic_ fashion.  This is probably most applicable to _Multiple Signatures_ since there's no other "nice" way of tying (per signature) meta-data to those:  https://cyberphone.github.io/doc/security/jcs.html#extensions

For _Simple Signatures_ the message body is (again IMO) usually more suited for meta-data that is not directly related to signature algorithms, keys, etc.

**Comment by mattcollier** [comment]
@cyberphone wrt your comment about picking the "best" property name the `https://w3id.org/security/v1` defines a `nonce` (just as an example) which can be used as shown in this example:
- http://json-ld.org/playground/#/gist/c1d7f6ebe2d90e570914d9eec998cac3

JSON-LD allows designers to use whatever property name they think is best and simply define the property they have chosen, `myTimeStamp` in this case, in an embedded or external context.  The signature is based on the `Expanded` output shown at the bottom of the playground example.

Does this address your concern?

**Comment by cyberphone** [comment]
@mattcollier I'm just looking at the LDS specification and the various mandatory and optional meta data properties which (MO) does not necessarily need to be a part of a signature scheme.   Maybe my NON-LD perspective distorts the picture?

**Comment by ChristopherA** [comment]
@cyberphone Note that you can add multiple signatures and that one of the other signature types is a secure timestamp using a merkle tree, which is much more reliable proof of existence than signature creation date. Other than that, you can always added a date to the signed data.

**Comment by cyberphone** [comment]
@ChristopherA FWIW, I'm exclusive looking at the LDS specification for "traditional" uses.  Merkle trees are (probably) great but do (AFAICT...) not make much sense for Verified Credentials and similar.

Personally, I would be _very cautious_ about overloading a scheme like LDS with "alien" stuff like Merkle trees, because that increases complexity dramatically, making all related projects suffer. When I (re)designed JCS I used the JOSE pieces that fitted (JWA+JWK), but totally ignored those that did not including the (nowadays _redundant_) dressing of data and headers in Base64. The result is a nimble scheme that is fairly simple to port and test on different platforms.  

Dropping "nice to have" meta data is one way of refining LDS, the other is moving _all_ crypto to JWS (or alternatively not use JWS).  The current crypto solution is a bit of a kludge (which is _fully understandable_ since it is a brand new design).

**Comment by ericelliott** [comment]
Is this field intended to represent the date it was signed, or the date the thing being signed was created?

If it's the former, would `dateSigned` work as a better property name?

**Comment by msporny** [comment]
@ericelliott - it's meant to represent the **datetime** when the signature was generated (accurate down to nanoseconds if using ISO8601 format). "created" is an vocabulary term that has 18 years of history behind it: http://dublincore.org/documents/dcmi-terms/#terms-created

While there may be more verbose names, the term we have right now has been in use for 18+ years to mean exactly what we're using it for.

**Comment by ericelliott** [comment]
@msporny Thanks for the clarification, and justification. I think we can use a custom context to apply a more specific property name in our apps and documentation to avoid confusion.

**Comment by msporny** [comment]
@ericelliott Yes, you always have the option of doing that since all of this is built on top of JSON-LD.
